pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -206,46 +211,57 @@ func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesRe
 	return nil
 }
 
-func (m *manager) AddContainer(p *v1.Pod, c *v1.Container, containerID string) error {
+func (m *manager) Allocate(p *v1.Pod, c *v1.Container) error {
 	m.Lock()
-	// Proactively remove CPUs from init containers that have already run.
-	// They are guaranteed to have run to completion before any other
-	// container is run.
-	for _, initContainer := range p.Spec.InitContainers {
-		if c.Name != initContainer.Name {
-			err := m.policyRemoveContainerByRef(string(p.UID), initContainer.Name)
-			if err != nil {
-				klog.Warningf("[cpumanager] unable to remove init container (pod: %s, container: %s, error: %v)", string(p.UID), initContainer.Name, err)
-			}
-		}
-	}
+	defer m.Unlock()
+
+	// TODO: Reuse CPUs allocated to init containers to mimic functionality
+	// from previous implementation below. Logic should probably be pushed into
+	// the policy now instead of doing it at this level.
+
+	//// Proactively remove CPUs from init containers that have already run.
+	//// They are guaranteed to have run to completion before any other
+	//// container is run.
+	//for _, initContainer := range p.Spec.InitContainers {
+	//	if c.Name != initContainer.Name {
+	//		err := m.policyRemoveContainerByRef(string(p.UID), initContainer.Name)
+	//		if err != nil {
+	//			klog.Warningf("[cpumanager] unable to remove init container (pod: %s, container: %s, error: %v)", string(p.UID), initContainer.Name, err)
+	//		}
+	//	}
+	//}
```

Comment:
@klueska @vpickard would it suffice to do a check for init container in `policy.AddContainer()`  and then remove that container's CPUs from the shared pool? 
So at the bottom of `policy.AddContainer()` we could call out to something like:
``````
func (p *staticPolicy) releaseInitContainerCPUs(s state.State, pod *v1.Pod, container *v1.Container) {
        for _, initContainer := range pod.Spec.InitContainers {
                if container.Name == initContainer.Name {
                        toRelease, _ := s.GetCPUSet(string(pod.UID), container.Name)
                        // Mutate the shared pool, adding released cpus.
                        s.SetDefaultCPUSet(s.GetDefaultCPUSet().Union(toRelease))
                }
        }
}

Reply:
I'd need to look closer at this, but is the idea to:
1) Unconditionally allocate CPUs to the container from the pool of available CPUs 
2) Check if the container we just allocated to is an init container
3) if it IS an init container, reset the pool of available CPUs to re-include the CPUs just assigned to the init container (but keep them assigned to the init container in the process).
4) If it is NOT an init container, just return (leaving the CPUs removed from the pool of available CPUs).

I think this would probably work, and it's actually fairly simple to understand.

In the process, it probably makes sense to rename the function to policy.Allocate() instead of policy.AddContainer() since it is (and actually always was) just doing the CPU allocation, and not the actual taskset assignment through the container runtime.

Reply:
Yes, this is the flow I was thinking of. It seems like the simplest solution. I will push the fix. 
I'll also rename the function as mentioned.

Reply:
I guess the problem in this solution is that the init container will not be running on exclusive CPUs even if it has Guaranteed QoS, so some other container from another Pod might end up using the same CPUs?

Reply:
if i follow this correctly, it should be fine because init containers are run in sequence.  i will review the remaining commits to see if this works out.